slam_toolbox/src/loop_closure_assistant.cpp

```diff
@@ -125,48 +125,93 @@ void LoopClosureAssistant::publishGraph()
 /*****************************************************************************/
 {
   interactive_server_->clear();
-  std::unordered_map<int, Eigen::Vector3d>* graph = solver_->getGraph();
+  karto::MapperGraph * graph = mapper_->GetGraph();
 
-  if (graph->size() == 0)
+  if (!graph || graph->GetVertices().empty())
   {
     return;
   }
 
-  ROS_DEBUG("Graph size: %i",(int)graph->size());
   bool interactive_mode = false;
   {
     boost::mutex::scoped_lock lock(interactive_mutex_);
     interactive_mode = interactive_mode_;
   }
 
-  visualization_msgs::MarkerArray marray;
-  visualization_msgs::Marker m = vis_utils::toMarker(map_frame_,
-    "slam_toolbox", 0.1);
+  // schedule all existing markers for removal
+  for (auto it = markers_.begin(); it != markers_.end(); ++it) {
+    it->second.action = visualization_msgs::Marker::DELETE;
+  }
+
+  visualization_msgs::Marker vertex_marker =
+    vis_utils::toVertexMarker(map_frame_, "slam_toolbox", 0.1);
 
-  for (ConstGraphIterator it = graph->begin(); it != graph->end(); ++it)
+  for (const auto & vertices_per_sensor : graph->GetVertices())
   {
-    m.id = it->first + 1;
-    m.pose.position.x = it->second(0);
-    m.pose.position.y = it->second(1);
+    for (const auto & id_and_vertex : vertices_per_sensor.second)
+    {
+      karto::LocalizedRangeScan * scan = id_and_vertex.second->GetObject();
+
+      vertex_marker.id = scan->GetUniqueId() + 1;
+      vertex_marker.pose.position.x = scan->GetCorrectedPose().GetX();
+      vertex_marker.pose.position.y = scan->GetCorrectedPose().GetY();
+
+      if (interactive_mode && enable_interactive_mode_)
+      {
+        visualization_msgs::InteractiveMarker int_marker =
+          vis_utils::toInteractiveMarker(vertex_marker, 0.3);
+        interactive_server_->insert(int_marker,
+          boost::bind(
+          &LoopClosureAssistant::processInteractiveFeedback,
+          this, _1));
+      }
+      else
+      {
+        markers_[vertex_marker.id] = vertex_marker;
+      }
+    }
+  }
 
-    if (interactive_mode && enable_interactive_mode_)
+  if (!interactive_mode)
+  {
+    visualization_msgs::Marker edge_marker =
+      vis_utils::toEdgeMarker(map_frame_, "slam_toolbox", 0.05);
+
+    for (auto * edge : graph->GetEdges()) {
+      karto::LocalizedRangeScan * source_scan = edge->GetSource()->GetObject();
+      karto::LocalizedRangeScan * target_scan = edge->GetTarget()->GetObject();
+      const karto::Pose2 source_pose = source_scan->GetCorrectedPose();
+      const karto::Pose2 target_pose = target_scan->GetCorrectedPose();
+
+      edge_marker.id =
+          (source_scan->GetUniqueId() << 16) + target_scan->GetUniqueId();
```

Comment:
what's what this naming?

Reply:
That's a best effort, arguably not elegant approach to deriving 32 bit edge IDs from 32 bit source and target vertices IDs. As long as vertices' unique IDs stay under 2^15 - 1 (IDs are signed for some reason), there won't be clashes. 

A hash function for a pair of integers may work too. Can't make guarantees about collision rates then.

Reply:
Couldn't you make vertices count from 1,2,3,4... while you make edges count down 2^15, 2^15-1, ...? Might make things easier than a hashing function

Reply:
Do you mean attaching a unique ID to edges? Those will collide with vertex IDs over time too. 

Issue here is that we need some function that relates graph vertices and edges with their markers if we want to remove (resp. update) the latter when the former are removed (resp. updated).

Reply:
@SteveMacenski PTAL at https://github.com/SteveMacenski/slam_toolbox/pull/467/commits/07961ed95217360e067a4505414077a1d4533890